accepted/single-file/design.md

```diff
@@ -66,225 +32,254 @@ Publishing to a single file can be triggered by adding the following property to
-* Setting the `PublishSingleFile`property causes the managed app, managed dependencies, platform-specific native dependencies, configurations, etc. (basically the contents of the publish directory when `dotnet publish` is run without setting the property) to be embedded within the native `apphost`. 
+When the `PublishSingleFile` property is set to true, 
 
-By default, the symbol files are not embedded within the single-file, but remain as separate files in the publish directory. 
-This includes both the IL `.pdb` file, and the native `.ni.pdb` / `app.guid.map` files generated by ready-to-run compiler.
-Setting the following property causes the symbol files to be included in the single-file.
+* `RuntimeIdentifier` must be defined. Single-file builds generate a native binary for the specific platform and architecture.
+* `UseAppHost` cannot be  set to `false`.  
+* If `TargetFramework` is
+  * `netcoreapp5.0`, single-file publish works as described in this document.
+  * `netcoreapp3.0`  or `netcoreapp3.1` single-file publish works as described [here](design_3_0.md).
+  * An earlier framework, causes a compilation error.
 
-```xml
-<PropertyGroup>
-    <IncludeSymbolsInSingleFile>true</IncludeSymbolsInSingleFile>
-</PropertyGroup>
-```
+Setting the `PublishSingleFile` property causes the managed app, runtime configuration files (`app.deps.json`, `app.runtimeconfig.json`), and managed binary dependencies to be embedded within the native `apphost`.  All managed binaries (IL and ready-to-run files) that would be written to the publish directory and any sub-directories are bundled with the apphost.
 
-Certain files can be explicitly excluded from being embedded in the single-file by setting following meta-data:
+All other files, including platform-specific native binaries and symbol files, are left alongside the app by default. However, the set of files left unbundled alongside the app is expected to be small, such as: data-files (ex: `appsettings.json`) and custom native binary dependencies of the application. Further details regarding the files left next to the app is discussed in the [Host build](#Host-Builds) section. 
 
-```xml
-<ExcludeFromSingleFile>true</ExcludeFromSingleFile>
-```
+### Optional Settings
 
-For example, to place some files in the publish directory but not bundle them in the single-file:
+| Property                        | Behavior when set to `true`                                  |
+| ------------------------------- | ------------------------------------------------------------ |
+| `IncludeContentInSingleFile`    | All published files (except symbol files) are embedded within the single-file bundle. |
+| `IncludeSymbolsInSingleFile`    | The symbol files (IL `.pdb` file, and the native `.ni.pdb` / `app.guid.map` files generated by ready-to-run compiler) are embedded within the symbol file. |
```

Comment:
This option is going to be useful only in combination with ExtractContentsFromSingleFile. Is that correct?

Or are we going to fix all tools that consume .ni.pdb, .map files, etc. to deal with them when they are embedded? It would be pretty hard to do, I think.

Reply:
Yes, the symbols are only useful when extracted. 

Similarly `IncludeContentInSingleFile` is only useful with `ExtractContentsFromSingleFile` unless they are read through the new APIs that were proposed. 

So, I have simplified the publish properties to remove `ExtractContentsFromSingleFile` option. Instead `IncludeContentInSingleFile` and `IncludeSymbolsInSingleFile` will imply extraction.

Reply:
> Yes, the symbols are only useful when extracted.

That is true for unmanaged symbols (.ni.dll/.map). For managed symbols, we can do the work to map them from the bundle for the stacktraces. Or we can just tell people to use .pdbs embedded into IL image if they want to have .pdbs for the single file apps (ie no special work required for single file).

For the unmanaged symbols, I suspect that there would be number of other changes required. For example, .map files contain absolute file offset, so you would need to rewrite them as part of the embedding.

Similar on Windows, the sampling profilers depend on the file to be mapped as proper Windows module. So to make them work, I believe you would need to merge all R2R .PDBs into one that matches the single-file module - or something like that.

I think this paragraph about symbols needs to have customer scenarios: What is and is not going to work, and why it is important. Right now it just talks about implementation details that is not good enough.

BTW: I am not convinced that we need to be embedding and extracting the unmanaged symbols (.map/.ni.pdbs). The scenarios for it sounds pretty marginal. However, we should make sure that the unmanaged symbols are still usable after embeding - that is the important part.

Reply:
I spoke with @mikem8361  about this. 

Mike said that the the `.map` files store physical offsets that are interpretted wrt the module load. Today, R2R assemblies are loaded on Linux via mapping the files and adjusting offsets -- the `.map` files are expected to work correctly in this case. For single-file apps, R2R assemblies will again be loaded via MappedImageLayout (albeit after mapping from a different offset within the single-file bundle). So, the `.map` files are expected to work OK.

Mike also said that the Windows profiler should be able to handle the `.ni.pdb` files without the need to merge them into one. But I don't know much details about the profiler.

I agree that embedding symbol files within the bundle is not an important use case. `IncludeSymbolsInSingleFile` was provided in .net core 3 based on request by a few customers -- so I carried forward the option. But nobody has asked for embedding the `.ni.pdb`/ `.map` files yet.

Reply:
Thank you for doing the homework on this. 

- `IncludeSymbolsInSingleFile` option is provided for compatibility with .NET 3 single-file mode. It extract files to disk.
- Embedded portable PDBs are the solution for proper no-extract-to-disk single file. This should work today already.
- Unless we hear somebody asking for it with a good backing scenario, we will not do anything for embedding the .ni.pdb/ .map files